src/connection.rs

```diff
@@ -706,19 +704,22 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Connection<T> {
             if let Some(w) = shared.reader.take() {
                 w.wake()
             }
-            if !is_finish
-                && shared.window == 0
-                && self.config.window_update_mode == WindowUpdateMode::OnReceive
-            {
-                shared.window = self.config.receive_window;
-                let frame = Frame::window_update(stream_id, self.config.receive_window);
-                return Action::Update(frame)
-            }
-        } else if !is_finish {
-            log::debug!("{}/{}: data for unknown stream", self.id, stream_id);
-            let mut header = Header::data(stream_id, 0);
-            header.rst();
-            return Action::Reset(Frame::new(header))
+            if matches!(self.config.window_update_mode, WindowUpdateMode::OnReceive) {
+                if let Some(credit) = shared.next_window_update() {
+                    shared.window += credit;
+                    let frame = Frame::window_update(stream_id, credit);
+                    return Action::Update(frame)
+                }
+            }
+        } else {
+            log::debug!("{}/{}: data for unknown stream, ignoring", self.id, stream_id);
+            // Previous implementations would send back a stream reset when receiving a data frame
+            // on an unknown stream. There are multiple benign scenarios that can lead to this
+            // happening and where a stream reset would not be appropriate. One such case would be
+            // when still in the process of sending out a frame closing the stream, while the stream
+            // data structure itself has already been garbage collected.
```

Comment:
```suggestion
            // We do not consider this a protocol violation and thus do not send a stream reset
            /// because we may still be processing pending `StreamCommand`s of this
            /// stream that were sent before it has been dropped and "garbage collected".
            /// Such a stream reset would interfere with the frames that still need to be
            /// sent, causing premature stream termination for the remote. 
```
Just a suggestion, I don't think it is necessary to refer to previous implementations (there is the git history). Feel free to ignore. If you take the suggestion, then it should be the same in all places, of course.

Reply:
Better for sure. Done with https://github.com/paritytech/yamux/pull/109/commits/55db98a5133b5c563ab56bb7b06338d57bdca168.